### ISSUE TITLE
When testing Phoenix Framework controllers, @endpoint must be set multiple times

### ISSUE THREAD

**Issue by mspanc** [sproutapp/pavlov#42]
I am using pavlov to test controllers in my phoenix app.

By default, phoenix generates `test/support/conn_case.ex` that sets `@endpoint MyApp.Endpoint` in `using` block in the following way:

```
defmodule MyApp.ConnCase do
  @moduledoc """
  This module defines the test case to be used by
  tests that require setting up a connection.

  Such tests rely on `Phoenix.ConnTest` and also
  imports other functionality to make it easier
  to build and query models.

  Finally, if the test case interacts with the database,
  it cannot be async. For this reason, every test runs
  inside a transaction which is reset at the beginning
  of the test unless the test case is marked as async.
  """

  use ExUnit.CaseTemplate

  using do
    quote do
      # Import conveniences for testing with connections
      use Phoenix.ConnTest

      alias MyApp.Repo
      import Ecto.Model
      import Ecto.Query, only: [from: 2]

      import MyApp.Router.Helpers

      # The default endpoint for testing
      @endpoint MyApp.Endpoint
    end
  end

  setup tags do
    unless tags[:async] do
      Ecto.Adapters.SQL.restart_test_transaction(MyApp.Repo, [])
    end

    :ok
  end
end
```

Thanks to this you have to set endpoint once.

Unfortunately, when I use contexts in pavlov-based tests it seems that I have to set `@endpoint` again and again in each of contexts. This will work:

```
defmodule MyApp.UserControllerTest do
  use MyApp.ConnCase, async: true
  use Pavlov.Case, async: true
  import Pavlov.Syntax.Expect

  let(:base_url) do
    "/api/rest/channel/input/stream/rtp" 
  end


  describe "POST (base_url)" do
    context "if no Authorization header was passed" do
      @endpoint MyApp.Endpoint
      
      it "returns status code of 401 Unauthorized" do 
        response = post conn(), base_url
        expect response.status |> to_eq 401
      end
    end
  end
end
```

But this won't:

```
defmodule MyApp.UserControllerTest do
  use MyApp.ConnCase, async: true
  use Pavlov.Case, async: true
  import Pavlov.Syntax.Expect

  let(:base_url) do
    "/api/rest/channel/input/stream/rtp" 
  end


  describe "POST (base_url)" do
    @endpoint MyApp.Endpoint

    context "if no Authorization header was passed" do
      it "returns status code of 401 Unauthorized" do 
        response = post conn(), base_url
        expect response.status |> to_eq 401
      end
    end
  end
end
```

or

```
defmodule MyApp.UserControllerTest do
  use MyApp.ConnCase, async: true
  use Pavlov.Case, async: true
  import Pavlov.Syntax.Expect

  let(:base_url) do
    "/api/rest/channel/input/stream/rtp" 
  end

  @endpoint MyApp.Endpoint

  describe "POST (base_url)" do
    context "if no Authorization header was passed" do
      it "returns status code of 401 Unauthorized" do 
        response = post conn(), base_url
        expect response.status |> to_eq 401
      end
    end
  end
end
```

It will end up with the following error:

```
     ** (RuntimeError) no @endpoint set in test case
     stacktrace:
       (phoenix) lib/phoenix/test/conn_test.ex:183: Phoenix.ConnTest.dispatch/5
       test/controllers/channel/input/stream/rtp_controller_test.exs:34
```

I guess that context/describe macros change scope in such a way that `@endpoint` is no longer accessible, but it's extremely unconvinient to define it multiple times and this breaks DRY principle.

Obviously you can say that this is not pavlov issue, but it's just wrong that Phoenix.ConnTest depends on `@endpoint` but well, I think it's pavlov that should somehow address this issue because module attributes are quite common in Elixir world.

**Comment by inf0rmer** [comment]
Hi! As you said, using `describe` or `context` dynamically defines a module using the text you provide for its name. The reason for this is so that things (such as a `let` or a`before`) in one of your scopes don't end up existing and running in another context.

A possible fix could be to redeclare every module attribute in every child scope, but unfortunately it seems to be impossible to get a list of all module attributes declared via the `@` macro (`__MODULE__.__info__(:attributes)` only returns `vsn`).

Another possible idea is to use a specific DSL for module attributes, that would then cache this value and declare it in all module scopes created by Pavlov. Something like:

```elixir
defmodule MyApp.UserControllerTest do
  use MyApp.ConnCase, async: true
  use Pavlov.Case, async: true
  import Pavlov.Syntax.Expect

  let(:base_url) do
    "/api/user" 
  end

  with_module_attribute :endpoint, MyApp.Endpoint

  describe "POST (base_url)" do
    context "if no Authorization header was passed" do
      it "returns status code of 401 Unauthorized" do 
        response = post conn(), base_url
        expect response.status |> to_eq 401
      end
    end
  end
end
```

An implementation of `with_module_attribute` would then just call `Module.put_attribute/3` to set the attribute. What do you think?

**Comment by mspanc** [comment]
Yes, I think it will do the job. Do you mean that I should implement this by myself or do you think about adding this to pavlov?

**Comment by inf0rmer** [comment]
I can implement it over the next few days, no problem. Of course, if you wish to contribute to the project you're more than welcome!






On Sun, Aug 2, 2015 at 1:14 PM -0700, "Marcin Lewandowski" <notifications@github.com> wrote:












Yes, I think it will do the job. Do you mean that I should implement this by myself or do you think about adding this to pavlov?

**Comment by mspanc** [comment]
This time I am chasing deadlines, so I am afraid I won't be able to help but generally speaking I'll be glad to support development of pavlov one day as it brings niceness of RSpec into Phoenix :)

**Comment by mspanc** [comment]
One thought: wouldn't it better to implement more general syntax? Something such as

```elixir
before(:scope) do
  SOME_WAY_TO_DETERMINE_MODULE_FOR_CURRENT_SCOPE |> Module.put_attribute :endpoint, MyApp.Endppoint
end
```

So another `before` callback that passes current scope and gives a chance to do any modifications.

I think it can be more elastic and won't bloat the pavlov API with function that just does one thing.

But if it's too complicated, it's better to have `with_module_attribute` soon than something that largely affects architecture later.

**Comment by inf0rmer** [comment]
`__MODULE__` is already an alias for the module in the current scope, but if you try setting an attribute in a `before` block you'll hit a runtime error:

```elixir
before do
  Module.put_attribute __MODULE__, :endpoint, MyApp.Endppoint
end

=> ** (ArgumentError) could not call put_attribute on module :"YOUR_MODULE_NAME" because it was already compiled
```

You cannot set attributes after a module is compiled (essentially after the `end` in the `defmodule` block) so, because `before` blocks run before the tests, it's basically impossible to set the attributes there.

Another option that just occurred to me would be to pass attributes in `describe/context`, which may feel more natural:

```elixir
describe "POST (base_url)", endpoint: MyApp.Endpoint, foo: "bar" do
  # "endpoint" and "foo" would be set to module attributes
end
```

Do you think this would look better?

**Comment by mspanc** [comment]
Definitely the latter solution (with passing attributes to the `describe/context`) looks much clearer. I would just suggest wrapping them into keyword lists, in case you wanted to add more attributes to `describe/context` with different meaning in the future.

So then it would have looked like:

```elixir
describe "POST (base_url)", module_attributes: [endpoint: MyApp.Endpoint, foo: "bar"] do
  # "endpoint" and "foo" would be set to module attributes
end
```

That leaves you some open path if you ever decide to create more attributes that can adjust `describe/context`.

The last question is, is it possible that it will automatically propagate into nested `describe/context`? So that

```elixir
describe "POST (base_url)", module_attributes: [endpoint: MyApp.Endpoint, foo: "bar"] do
  # "endpoint" and "foo" would be set to module attributes
  context "if something happens" do
    # "endpoint" and "foo" would be set to module attributes, too, unless context has it's own list of module_attributes that will override parent one

  end
end
```

?

**Comment by inf0rmer** [comment]
Yeah, the idea is that they would be made available in child contexts as well. Possibly even by merging the keyword lists, so that you don't overwrite previous definitions unless you explicitly set a new value. 

I'll work on this proposal then, let's see if a release sometime this week is possible, together with the work already merged to implement support for `subject`. 

Thanks for the help in designing the feature!

**Comment by b-murphy** [comment]
@inf0rmer did you have any luck with this?

I just wanted to throw my 2p into a possible spec. having read the issue above & playing around a little (fairly new to Elixir, from Ruby) I landed up using this as my solution.

```
describe "create" do
  use MyApp.ConnCase, async: true

  it "makes a request" do
   response = post(conn(), "somewhere")
   expect(response.status) |> to_eq(201)
  end
end
```

This allowed me to bring the other custom matchers i'd previously imported into ConnCase as i'd previously build the test cases with the default ExUnit. Not the best solution, but make it easier to port over to Pavlov.

So i was going to say could there be some consideration to the inclusion of modules in a more RSpec style syntax where we would use simple `types` to delegate the `describe` modules spec on run time?

```
describe "create", :controller do
  it "makes a request" do
   response = post(conn(), "somewhere")
   expect(response.status) |> to_eq(201)
  end
end
```

This might be a bigger move as it would require Pavlov to have a way of defining these as modules.

Let me know if i can help out, Pavlov will be a big thing for our team if we can get this nailed :-)
